Xo update

nick-drozd opened the pull request:
The xo spec hasn't been written yet, but the xo TP test is intended to show what will go into the spec. Note that the xo message factory also contains information that will go into the spec, since it determines the format of xo state data.

This PR makes several changes to xo message formats: 1) the format of transactions has been changed from `(name, action, space)` to `(action, name, space)` (the latter being how transactions are submitted at the CLI); 2) xo addresses are like intkey addresses -- the first six characters of the hash of the family name plus the last 64 characters of the hash of the game name; and 3) the data stored at an address is now a JSON string in the following format: `{"game_name": "board,state,player_1,player_2"}`.

The Python xo implementation has been rewritten to more or less match the documentation, and a few errors have been fixed in the JS implementation.

dplumb94 commented:
3) should be: 

> the data stored at an address is now a JSON string in the following format: {"**game_name**": "board,state,player_1,player_2"}.

boydjohnson commented:
The change to storing json objects in state doesn't seem like it will work. So when/if a hash collision happens, there will be two keys in the json object and since json objects are unordered (http://www.rfc-editor.org/rfc/rfc7159.txt) we can't be certain the order of bytes stored in state. This will cause two different validators to get a merkle root that is different for state changes that are the same. bson is a binary format that is somewhat like json and says that bson "documents" are ordered key:value pairs. http://bsonspec.org/spec.html

dplumb94 commented on sdk/examples/xo_python/sawtooth_xo/processor/handler.py:
```diff
@@ -214,66 +217,157 @@ def _store_game_data(state_store, name, board, state, player_1, player_2):
+
+
+def _is_win(board, mark):
+    '''
```
Very clever, but I'm wondering if we should simplify this? Since the goal of XO is to show an example of how a simple transaction processor is written, extra complexity in the game logic may be counterproductive. We would also have to implement this in the other languages.

nick-drozd replied:
To me, the use of having a simple example like xo is to show how a transaction processor handles transactions and state data. The basic model that both xo and intkey follow is: get the transaction data, get the state data, do something, put the result back into state. From the perspective of showing off the system, it's relatively uninteresting what happens in the "do something" step, and the current documentation says so. It might even be more interesting if the implementations do differ, so long as they all produce the same output.

nick-drozd commented:
@boydjohnson raises a good point, so it's back the drawing board with the state format. I've pulled that commit, but the rest of this PR is unaffected.

nick-drozd commented:
Okay, hopefully this is the last update. @vaporos suggested that we use the following state format: `a_game,board,state,p1,p2|b_game,board,state,p1,p2|...`. Although it's a little unwieldy to implement, it has the benefit of not requiring any libraries or getting involved with any JSON baggage. On the other hand, this format has the disadvantage of potentially causing confusion regarding the use of csv in both the transaction format and the state format -- one might think that they are related, when really they can be completely independent choices.

Given this format, game names can't contain `,` or `|`. (Really, this restriction only applies to `|`, as an implementation would typically screen out names containing `,` when trying to decode the transaction.)

aludvik commented on sdk/examples/xo_python/sawtooth_xo/xo_client.py:
```diff
@@ -68,26 +68,38 @@ def take(self, name, space):
         try:
-            return base64.b64decode(yaml.safe_load(result)["data"])
-
+            return self.list()[name]
```
This downloads all of XO state to show one game, rather than construct the address to retrieve only one game, which is inefficient. I think it would be better to show how to query for only one address here.

aludvik commented on bin/run_tests:
```diff
@@ -302,6 +302,9 @@ test_xo() {
     run_docker_test xo-javascript-smoke -s integration_test
     copy_coverage .coverage.xo-javascript-smoke
+    # all languages
+    run_docker_test xo-all-lang-smoke
```
I think this could just be called "xo-all-lang".

nick-drozd commented on bin/run_tests:
```diff
@@ -288,6 +291,19 @@ test_deployment() {
     run_docker_test systemd_services -s integration_test
 }
 
+test_xo() {
```
If changes are made to the whole xo family (for instance, changing the namespace), it's convenient to be able to run all the xo tests at once.

The main `run_tests` module list, however, doesn't include the xo module:
```
MODULE_LIST="
    signing
    cli
    rest_api
    poet
    config
    validator
    go_sdk
    java_sdk
    javascript_sdk
    python_sdk
    integration
"
```
Someone working on xo can run `run_tests -m xo`, but nobody else needs to worry about it. The PoET unit tests can already be run in this way.

nick-drozd commented:
@aludvik I changed `two_families.yaml` to `two-families.yaml` to match the rest of the yaml files, and I changed `xo-smoke.yaml` to `xo-python-smoke.yaml` to match the rest of the smoke test yaml file. Neither of these changes are interesting enough to warrant separate commits, so I changed them when I changed the actual test files.

I changed the transaction format because it's confusing to have `take game 5` at the command line get changed into `game,take,5` when the transaction is sent.

Xo Javascript was organized somewhat haphazardly, so I moved things around to make it easier to change what needed to be changed. Think of it as a very mild refactor.

The rest client changes make it easier to verify state, and since the xo smoke test was updated to verify state, I included those changes in that commit. I've updated that commit slightly to make the change from `_base_url` to `url` clearer (it's convenient to be able to tell what url a client is pointed at).

vaporos commented on sdk/examples/xo_python/sawtooth_xo/xo_client.py:
```diff
@@ -30,8 +30,21 @@
+def _hash_name(name):
+    return hashlib.sha512(name.encode('utf-8')).hexdigest()
+
+FAMILY_NAME = 'xo'
```
Hard-coding FAMILY_NAME and XO_NAMESPACE would make it more difficult to re-use this code to run xo games against another namespace.  In previous iterations, games could be played using an arbitrary namespace.

vaporos commented on sdk/examples/xo_python/sawtooth_xo/xo_client.py:
```diff
@@ -49,45 +62,47 @@ def __init__(self, base_url, keyfile):
 
     def list(self):
-        xo_prefix = self._get_prefix()
+        state = yaml.safe_load(
```
Should handle invalid YAML coming out of self._send_request() here.

vaporos commented on sdk/examples/xo_python/sawtooth_xo/xo_client.py:
```diff
@@ -164,3 +179,7 @@ def _create_batch_list(self, transactions):
         return BatchList(batches=[batch])
+
+
+def _sha512(data):
```
Is this used?

vaporos commented on sdk/examples/xo_python/sawtooth_xo/xo_message_factory.py:
```diff
@@ -13,82 +13,110 @@
-            family_name="xo",
-            family_version="1.0",
-            namespace="",
+            encoding='csv-utf8',
```
Many unnecessary quote or whitespace changes in this block.

vaporos commented on sdk/examples/xo_python/tests/test_tp_xo.py:
```diff
@@ -13,85 +13,320 @@
+
+        self.expect_invalid()
+
+    @skip('JS throws InternalError instead of InvalidTransaction')
```
Let's exclude the test instead of skipping it, and add a JIRA for adding it.
